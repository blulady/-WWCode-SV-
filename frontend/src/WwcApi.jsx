import axios from "axios";

const BASE_URL =
  process.env.REACT_APP_API_URL ||
  "https://wwcode-chtools-api-dev.herokuapp.com/api";

const getConfig = () => {
  let { access } = JSON.parse(sessionStorage.getItem("token"));
  return {
    Authorization: `Bearer ${access}`,
    "Content-Type": "application/json",
  };
};

axios.interceptors.response.use(
  (response) => {
    return response;
  },
  function (error) {
    
    const {detail,code, error: err} = error.response.data
    console.log(err);
    console.log('Interceptor error:',detail,':',code)
     
    const originalRequest = error.config;
    
    if (error.response.status === 401 && !originalRequest._retry && code==='token_not_valid'){
      // Access Token expired, get new token using the existing Refresh token
      if (sessionStorage.getItem("token") && detail==='Given token not valid for any token type'){
          originalRequest._retry = true;
          const { refresh } = JSON.parse(sessionStorage.getItem("token"));
          return axios
            .post(`${BASE_URL}/login/refresh`, { refresh: refresh })
            .then((res) => {
              if (res.status === 200) {
                console.log(res);
                let { access, refresh } = res.data;
                let token = { access, refresh };
                sessionStorage.setItem("token", JSON.stringify(token));
                console.log("Access token has been refreshed");
                originalRequest.headers = getConfig();
                return axios(originalRequest);
              } 
            });   
          } 
          // Refresh Token expired. clean up the tokens stored in sessionStorage
          if (detail==='Token is invalid or expired'){
            console.log("Refresh token expired, cleanup tokens and navigate to Login page")
            sessionStorage.removeItem("token");
            sessionStorage.removeItem("user");
            return window.location.href = `${BASE_URL}/login`;
          }   
    }
    return Promise.reject(error);
  }
);

class WwcApi {
  static async login(data) {
    return await axios.post(`${BASE_URL}/login`, data, {
      headers: { "content-type": "application/json" },
    });
  }

  static async sendResetEmail(email) {
    return await axios.post(
      `${BASE_URL}/user/reset_password/request/`,
      { email: email },
      {
        headers: { "content-type": "application/json" },
      }
    );
  }

  static async resetPassword(data){
    return await axios.patch(`${BASE_URL}/user/reset_password/confirm/`, data, {
      headers: { "Content-Type": "application/json" }});
  }

  static async logout() {
    return await axios.post(
      `${BASE_URL}/logout/`,
      {},
      {
        headers: getConfig(),
      }
    );
  }

  static async getMembers(orderBy, search) {
    let url = `${BASE_URL}/users/`;
    if (orderBy) {
      url += "?ordering=" + orderBy;
    }
    if (search) {
      url += (orderBy ? "&" : "?") + "search=" + search
    }
    let res = await axios.get(url, {
      headers: getConfig(),
    });
    return res.data;
  }

  static async getMember(userId) {
    let res = await axios.get(`${BASE_URL}/user/${userId}`, {
      headers: getConfig(),
    });
    return res.data;
  }

  static async createMember(data) {
    return await axios.post(`${BASE_URL}/user/create/`, data, {
      headers: getConfig(),
    });
  }

  static async register(data) {
    return await axios.post(`${BASE_URL}/user/activate/`, data);
  }

  static async getTeams() {
    let res = await axios.get(`${BASE_URL}/teams/`,{
      headers: getConfig(), 
    });
    return res.data;
  }

  static async editMember(data, userId) {
    return await axios.post(`${BASE_URL}/user/edit/${userId}`, data,{
      headers: getConfig(),
    });
  }

  static async getVolunteerResources(slug) {
    let res = await axios.get(`${BASE_URL}/resources/${slug}`, {
      headers: getConfig(),
    });
    return res.data;
  }

  static async changeMemberStatus(userId,data){
    return await axios.post(`${BASE_URL}/user/edit/${userId}/status`, data, {
      headers: getConfig(),
    });
  }

  static async deleteMemberRole(userId,data){
    return await axios.delete(`${BASE_URL}/user/edit/${userId}/role`, data, {
      headers: getConfig(),
    });
  }
}
export default WwcApi;
