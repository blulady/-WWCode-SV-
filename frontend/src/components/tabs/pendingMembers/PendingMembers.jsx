import React from "react";
import { useEffect } from "react";
import { useState } from "react";
import { isBrowser } from "react-device-detect";
import { useNavigate } from "react-router-dom";
import WwcApi from "../../../WwcApi";
import PendingMemberList from "./PendingMemberList";
import PendingMemberTable from "./PendingMemberTable";
import MessageBox from "../../messagebox/MessageBox";
import {
  ERROR_TEAM_MEMBERS_UNABLE_TO_LOAD,
  ERROR_REQUEST_MESSAGE,
} from "../../../Messages";
import styles from "./PendingMembers.module.css";

const PendingMembers = () => {
  const [users, setUsers] = useState([]);
  const [apiError, setApiError] = useState(null);
  const [showMessage, setShowMessage] = useState(false);
  const navigate = useNavigate();

  useEffect(() => {
    getInvitees();
  }, []);

  const getInvitees = async () => {
    try {
      let _users = await WwcApi.getInvitees();
      setUsers(_users);
    } catch (error) {
      setApiError(ERROR_TEAM_MEMBERS_UNABLE_TO_LOAD.replace("{0}", ""));
      console.log(error);
    }
  };

    const goToAddMember = () => {
        navigate("/member/add",{
            state: { pageId: "chapter", pending: true }
          });
    };

  const resendInvite = async (currentUser) => {
    try {
      await WwcApi.resendInvite(currentUser);
      await getInvitees();
      setShowMessage(true);
    } catch (error) {
      setApiError(ERROR_REQUEST_MESSAGE);
      console.log(error);
    }
  };

  const handleDeleteMember = async (currentUserId) => {
    const temp = [...users];
    try {
      await WwcApi.deleteInvitees(currentUserId)
      const filteredMembers = temp.filter(
        (member) => member.id !== currentUserId
      );
      setUsers(filteredMembers);
    } catch (err) {
      setUsers(temp);
      setApiError(ERROR_REQUEST_MESSAGE);
      console.error(err)
    }
  };

  const renderTable = () => {
    return (
      <PendingMemberTable
        users={users}
        onDeleteMember={handleDeleteMember}
        onResendInvite={resendInvite}
      />
    );
  };
  const renderList = () => {
    return (
      <PendingMemberList
        users={users}
        onResendInvite={resendInvite}
        onDeleteMember={handleDeleteMember}
      ></PendingMemberList>
    );
  };

  return (
    <div className={styles["pending-members-container"]}>
      <div className="d-flex justify-content-end mb-2 mb-md-5">
        <button
          type="button"
          className="wwc-action-button"
          onClick={goToAddMember}
        >
          + Add Member
        </button>
      </div>
      {showMessage && (
        <div className="d-flex justify-content-center">
          <MessageBox
            type="Success"
            title="Success!"
            message="New registration link has been sent."
          ></MessageBox>
        </div>
      )}
      {apiError && (
        <div className="d-flex justify-content-center">
          <MessageBox
            type="Error"
            title="Sorry!"
            message={apiError}
          ></MessageBox>
        </div>
      )}
      {users.length ? (
        isBrowser ? (
          renderTable()
        ) : (
          renderList()
        )
      ) : (
        <div className={styles["no-users-msg"]}>No invitees to display</div>
      )}
    </div>
  );
};

export default PendingMembers;
