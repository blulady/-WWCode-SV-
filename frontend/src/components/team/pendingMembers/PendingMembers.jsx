import React from "react";
import { useEffect } from "react";
import { useState } from "react";
import { isBrowser } from "react-device-detect";
import { useNavigate } from "react-router-dom";
import WwcApi from "../../../WwcApi";
import PendingMemberList from "./PendingMemberList";
import PendingMemberTable from "./PendingMemberTable";
import ModalDialog from "../../common/ModalDialog";
import MessageBox from "../../messagebox/MessageBox";
import {
  ERROR_TEAM_MEMBERS_UNABLE_TO_LOAD,
  ERROR_REQUEST_MESSAGE,
} from "../../../Messages";
import styles from "./PendingMembers.module.css";

const PendingMembers = (props) => {
  const [users, setUsers] = useState([]);
  const [apiError, setApiError] = useState(null);
  const [showMessage, setShowMessage] = useState(false);
  const [currentUser, setCurrentUser] = useState(null);
  const navigate = useNavigate();

  const renderTable = () => {
    return (
      <PendingMemberTable
        users={users}
        onDeleteMember={handelDeleteMember}
        onResendInvite={resendInvite}
      />
    );
  };
  const renderList = () => {
    return (
      <PendingMemberList
        users={users}
        target="#resendConfirmationDialog"
        targetDelete="#deletePendingMemberDialog"
      ></PendingMemberList>
    );
  };

  const getInvitees = async () => {
    try {
      let _users = await WwcApi.getInvitees();
      setUsers(_users);
    } catch (error) {
      setApiError(ERROR_TEAM_MEMBERS_UNABLE_TO_LOAD.replace("{0}", ""));
      console.log(error);
    }
  };

    const goToAddMember = () => {
        navigate("/member/add",{
            state: { teamId: 0, pending: true }
          });
    };

  const resendInvite = async (currentUser) => {
    try {
      await WwcApi.resendInvite(currentUser);
      await getInvitees();
      setShowMessage(true);
    } catch (error) {
      setApiError(ERROR_REQUEST_MESSAGE);
      console.log(error);
    }
  };

  const handelDeleteMember = (currentUserId) => {
    const temp = [...users];
    const filteredMembers = temp.filter((member) => member.id !== currentUserId);
    setUsers(filteredMembers);
    WwcApi.deleteInvitees(currentUserId).catch((err) => {
      setUsers(temp);
      setApiError(ERROR_REQUEST_MESSAGE);
    });
  };

  const onOpeningModalDialog = (target) => {
    if (target) {
      const user = target.getAttribute("data-bs-user");
      setCurrentUser(user);
    }
  };

  useEffect(() => {
    getInvitees();
  }, []);

  return (
    <div className={styles["pending-members-container"]}>
      <div className="d-flex justify-content-end mb-2 mb-md-5">
        <button
          type="button"
          className="wwc-action-button"
          onClick={goToAddMember}
        >
          + Add Member
        </button>
      </div>
      {showMessage && (
        <div className="d-flex justify-content-center">
          <MessageBox
            type="Success"
            title="Success!"
            message="New registration link has been sent."
          ></MessageBox>
        </div>
      )}
      {apiError && (
        <div className="d-flex justify-content-center">
          <MessageBox
            type="Error"
            title="Sorry!"
            message={apiError}
          ></MessageBox>
        </div>
      )}
      {users.length ? (
        isBrowser ? (
          renderTable()
        ) : (
          renderList()
        )
      ) : (
        <div className={styles["no-users-msg"]}>No invitees to display</div>
      )}
    </div>
  );
};

export default PendingMembers;
