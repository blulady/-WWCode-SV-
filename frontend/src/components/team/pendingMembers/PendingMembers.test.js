import React from "react";
import { render, act } from "@testing-library/react";
import WwcApi from "../../../WwcApi";
import PendingMembers from "./PendingMembers";

jest.mock("../../../WwcApi", () => {
  return {
    ...jest.requireActual("../../../WwcApi"),
    getInvitees: jest.fn(),
  };
});

const mockNavigation = jest.fn();
jest.mock('react-router-dom', () => {
  const ActualReactRouterDom = jest.requireActual('react-router-dom');
  return {
    ...ActualReactRouterDom,
    useNavigate: () => mockNavigation
  }
});


describe("PendingMembers Component Validation Tests", () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  test("PendingMembers component is rendering", async () => {
    // Mock GET resources
    WwcApi.getInvitees.mockImplementation(async () => {
      return await Promise.resolve(
        [{
          id: 1,
          email: "abc@example.com",
          role_name: "volunteer",
          status: "invited"
        }]
      );
    });

    let res;
    await act(async () => {
      res = render(
        <PendingMembers />
      );
    });
    expect(res.container).toMatchSnapshot();

  });

  test("PendingMembers component renders no invitess message", async () => {
    // Mock GET resources
    WwcApi.getInvitees.mockImplementation(async () => {
      return await Promise.resolve([]);
    });

    let res;
    await act(async () => {
      res = render(
        <PendingMembers />
      );
    });

    expect(res.getByText("No invitees to display")).toBeTruthy();
  });

  test('PedingMembers component: Error while getting invitees', async () => {
    WwcApi.getInvitees.mockImplementation(async () => {
      throw new Error("No invitees");
    });

    let res;
    await act(async () => {
      res = render(
        <PendingMembers />
      );
    });

    expect(res.getByTestId('message-box')).toBeInTheDocument();
  });
});
