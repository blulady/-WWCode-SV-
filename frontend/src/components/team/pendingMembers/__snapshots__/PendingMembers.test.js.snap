// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PendingMembers Component Validation Tests PendingMembers component is rendering 1`] = `
<div>
  <div
    class="pending-members-container"
  >
    <div
      class="d-flex justify-content-end mb-2 mb-md-5"
    >
      <button
        class="wwc-action-button"
        type="button"
      >
        + Add Member
      </button>
    </div>
    <table
      class="pending-members-table"
    >
      <thead>
        <tr>
          <th />
          <th>
            Email
          </th>
          <th>
            Role
          </th>
          <th>
            Status
          </th>
          <th>
            Resend Invite
          </th>
          <th>
            Actions
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            1
          </td>
          <td>
            abc@example.com
          </td>
          <td
            class="wwc-text-capitalize"
          >
            volunteer
          </td>
          <td
            class="wwc-text-capitalize"
          >
            invited
          </td>
          <td>
            <button
              class="invite-button"
              data-bs-target="#resendConfirmationDialog"
              data-bs-toggle="modal"
              data-bs-user="1"
              type="button"
            >
              Resend Invite
            </button>
          </td>
          <td>
            <button
              class="delete icon"
            />
          </td>
        </tr>
      </tbody>
    </table>
    <div
      class="modal"
      id="resendConfirmationDialog"
      role="dialog"
      tabindex="-1"
    >
      <div
        class="modal-dialog modal-dialog-centered"
      >
        <div
          class="modal-content"
        >
          <div
            class="modal-header font-weight-bold"
          >
            <header
              class="text-center"
            >
              <div
                class="modal-title-text"
              >
                Are you sure?
              </div>
            </header>
          </div>
          <div
            class="modal-body"
          >
            <div
              class="mb-5"
            >
              Are you sure you want to resend the registration link?
            </div>
            <div
              class="text-center"
            >
              <button
                class="btn btn cancel-btn dialog-button mr-3"
                data-bs-dismiss="modal"
              >
                Cancel
              </button>
              <button
                class="btn btn confirm-btn dialog-button submit-button"
                data-bs-dismiss="modal"
                type="submit"
              >
                Confirm
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
